server/conn.go

```diff
@@ -0,0 +1,132 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package server contains business logic of working with pmm-managed.
+package server
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/percona/pmm/api/agent"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// Conn contains business logic of communication with pmm-managed.
+type Conn struct {
+	stream      agent.Agent_ConnectClient
+	lastID      uint32
+	l           *logrus.Entry
+	rw          sync.RWMutex
+	subscribers map[uint32]chan *agent.ServerMessage
+	requestChan chan *agent.ServerMessage
+}
+
+// NewConn starts goroutine to dispatch messages from server and returns new Conn object
+func NewConn(serverAddress string, stream agent.Agent_ConnectClient) *Conn {
+	conn := &Conn{
+		stream:      stream,
+		l:           logrus.WithField("server-address", serverAddress),
+		subscribers: make(map[uint32]chan *agent.ServerMessage),
+		requestChan: make(chan *agent.ServerMessage),
+	}
+	// create goroutine to dispatch messages
+	go conn.startMessageDispatcher()
+	return conn
+}
+
+// SendAndRecv sends requests to server and waits for response
+func (c *Conn) SendAndRecv(toServer agent.AgentMessagePayload) (*agent.ServerMessage, error) {
+	id := atomic.AddUint32(&c.lastID, 1)
+	agentMessage := &agent.AgentMessage{
+		Id:      id,
+		Payload: toServer,
+	}
+	c.l.Debugf("Send: %s.", agentMessage)
+	if err := c.stream.Send(agentMessage); err != nil {
+		return nil, errors.Wrap(err, "failed to send message to agent")
+	}
+
+	agentChan := make(chan *agent.ServerMessage, 1)
+
+	c.addSubscriber(id, agentChan)
+
+	serverMessage := <-agentChan
+	c.l.Debugf("Recv: %s.", serverMessage)
+
+	c.removeSubscriber(id, agentChan)
+	close(agentChan)
```

Comment:
In almost every situation the sender should close the channel. Closing a channel is send-like operation.
I think the sender should send a message and close the channel on the normal path, and just close it on the error path.
The receiver gets a single message. If it is `nil`, that's error and stream should be closed. Or it may be already closed by the sender, so receiver should just exit.